libc/dirent/closedir.c

```diff
@@ -0,0 +1,157 @@
+#include <time.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <dirent.h>
+#include <pthread.h>
+#include <signal.h>
+
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include <unity_fixture.h>
+
+#include "common.h"
+
+#define MAIN_DIR          "tested_files_closedir"
+#define INTERRUPT_DIR_NUM 256
+
+
+void close_dirs(void)
+{
+	for (int i = 0; i < INTERRUPT_DIR_NUM; ++i) {
+		char name[40];
+		sprintf(name, MAIN_DIR "/%d", i);
+		closedir(name);
+	}
+}
+
+TEST_GROUP(closedir);
+
+
+TEST_SETUP(closedir)
+{
+	mkdir(MAIN_DIR, 0777);
+	mkdir(MAIN_DIR "/dir1", 0777);
+}
+
+
+TEST_TEAR_DOWN(closedir)
+{
+	errno = 0;
+	rmdir(MAIN_DIR "/dir1");
+	rmdir(MAIN_DIR);
+}
+
+
+TEST(closedir, closing_opened_directory_normal)
+{
+	errno = 0;
+	DIR *dp = opendir(MAIN_DIR "/dir1");
+
+	TEST_ASSERT_NOT_NULL(dp);
+	TEST_ASSERT_EQUAL(0, closedir(dp));
+	TEST_ASSERT_EQUAL(0, errno);
+}
+
+
+TEST(closedir, closing_closed_directory)
+{
+	DIR *dirPtr = opendir(MAIN_DIR "/dir1");
+	errno = 0;
+
+	closedir(dirPtr);
+	TEST_ASSERT_EQUAL(-1, closedir(dirPtr));
+	TEST_ASSERT_EQUAL(EBADF, errno);
+}
+
+
+TEST(closedir, interrupt_with_sigkill)
+{
+	errno = 0;
+
+	DIR *dirs[INTERRUPT_DIR_NUM];
+	char *name[8];
+
+	for (int i = 0; i < INTERRUPT_DIR_NUM; ++i) {
+		sprintf(name, MAIN_DIR "/%d", i);
+		dirs[i] = opendir(name);
+	}
+
+	pid_t childID = fork();
+	if (childID) {
+		for (int i = 0; i < INTERRUPT_DIR_NUM; ++i) {
+			closedir(dirs[i]);
+		}
+	}
+
+	kill(childID, SIGKILL);
```

Comment:
`SIGKILL` also kills the parent - that's why it won't work. I am not sure whether this requirement is valid on our system and possible to test - to be discussed. Notice that it's not strictly required on POSIX as word `may` is used: `The closedir() function may fail if:`. For now you can leave those cases commented out.